### PR TITLE
[3.5] Update NavigationServer backport

### LINKED ISSUE #28235: Navigation2D#get_simple_path doesn't behave properly when using per-tile navigation polygons
**Godot version:**

3.1.stable.official

**OS/device including version:**

Windows 10 (10.0.17134 Build 17134), but happens on other systems too.

**Issue description:**

When adding `NavigationPolygonInstance`s that entirely cover tiles of a tileset, then covering a `Tilemap` with said tiles, using `Navigation2D` and its `get_simple_path` method produces unwanted behavior (unoptimal paths that include more points than needed):
https://i.imgur.com/yn4hegV.gif
https://streamable.com/47nzc

The actual result should instead be the same as what happens when covering the whole navigable area with a single `NavigationPolygonInstance`:
https://streamable.com/exk7u

**Steps to reproduce:**

Note that these steps can be reproduced either with the 3.1's new tileset editor, or using the old-fashioned way (scene converted to tileset).

1. Create a tileset resource with one single tile (to keep things simple).
2. Add a `NavigationPolygonInstance` covering exactly that whole tile.
3. Create a scene with a `Navigation2D` node that includes a `TileMap` node referencing that tileset. Cover the tilemap with the tile.
4. Add a sprite node that will serve as the moving object and place it somewhere on the map.
5. Use `get_simple_path` (with `optimize` to `true`) to compute a path between the sprite and a position far away (ideally, diagonally).
6. The resulted path will contain unwanted steps, even though there is no obstacle and the whole area is supposed to be navigable.

**Minimal reproduction project:**

[weird_pathfinding.zip](https://github.com/godotengine/godot/files/3100180/weird_pathfinding.zip)

**Notes:**

- This was initially posted as a question [here](https://godotengine.org/qa/43645/simple-navigation2d-pathfinding-issue) and someone agreed it looked like a bug (even remade his own minimal projects to find out it behaved slightly differently in 3.0.6).
- If you take [KidsCanCode's recent YouTube tutorial](https://www.youtube.com/watch?v=OYn49ghh9k0) (that you can [download from here](https://github.com/kidscancode/godot_tilemap_navigation/tree/master/tilemap_navigation_3.1_end)) and simply fill the "holes" on the tilemap with the given tile, it will occur as well (that's actually what I did for the streamable.com links above). I talked a bit with KidsCanCode on Discord and he thought that seemed suspicious.
- I'm still very much a Godot beginner, so sorry if I missed anything obvious.

### LINKED ISSUE #59182: Navigation2D (Godot 3.5 beta2) - Error: "sync: Attempted to merge a navigation mesh triangle edge with another already-merged edge", not present in Godot 3.4.3
### Godot version

3.5 beta 2

### System information

Windows 7 and 10, GLES2 and GLES 3

### Issue description

Bug submited in https://github.com/godotengine/godot/issues/57254 and supposedly fixed by https://github.com/godotengine/godot/pull/56879 and https://github.com/godotengine/godot/pull/56600 in still present in Navigation2D. Error spams in console, but NavigationPolygon is created and working. Error spams with polygons with hundreds of points, close to each other, when triangulating the big polygon by code. I attach reproduction project.

Error message:

```
E 0:00:00.999   sync: Attempted to merge a navigation mesh triangle edge with another already-merged edge. Either the Navigation's `cell_size` is different from the one used to generate the navigation mesh or `detail/sample_max_error` is too small. This will cause navigation problem.
  <C++ Source>  modules/navigation/nav_map.cpp:605 @ sync()
```

This is not a breaking bug per se, but a inconvenient spam in console message.

Other significant change since new navigation system is that you can´t use Navigation2D in the **_ready()** function when you are building polygons from code, and you have to workaround with the first frame in **_physics_process** to handle "Navigation2D.get_simple_path()". I made the changes in my project because i want to migrate to 3.5, but is a breaking change that should be documented.

### Steps to reproduce

Run attached project in 3.4.3 and 3.5 beta 2 and see output console.

### Minimal reproduction project

[NavigationMeshBug.zip](https://github.com/godotengine/godot/files/8257816/NavigationMeshBug.zip)

### PR DESCRIPTION
**Pull request by smix8**
Backports features and bugfixes from current Godot 4.0 to 3.5 and brings functions and codebase of both version largely in sync to make tutorials more compatible and future backports easier.

Fixes #59182
Fixes #28235

To my suprise I encountered fixes and additions made to the 3.5 branch only that were never added to master (like sometimes even announced but never done). I tried my best to add/keep them and will add those that still make sense to 4.0 in a dedicated pr later.

In particular this backports adds features and fixes from:

#47133 Make 2D navigation consider scale
#47024 Allow Navigation to be more flexible. This pr helps a ton with pathfinding and edge merging issues in 3.5 but also makes some 3.5 fixes invalid or at least they require a second look and if they still have purpose maybe reimplementation.
#46786 Implements the Navigation layers but does not remove the Navigation/Navigation2D nodes
#61800 Fix GridMap Navigation transform changes and debug
#61739 Add NavigationRegion costs for 2D and 3D pathfinding
#61589 Fix NavigationAgent reparent issues

### **Compatibility**

**NavigationPolygonInstance, NavigationMeshInstance and all NavigationAgents**
(not obstacles as they require more fixes in 4.0 as well and will get their own pr)

Compatibility for the old Navigation/Navigation2D nodes is available. e.g. if such nodes are present child navigation related nodes connect to the navigation map of navigation node (which is the old behavior) else they connect to the default navigation map of the world/world2d. This also means that all navigation related nodes can now be used without a Navigation/Navigation2D node parent as they will use the default map.

**TileMap and GridMap**

The TileMap and GridMap have a new `bake_navigation` property that must be set to use build-in navigation with the default navigation map. Previously this was toggled automatically when a Navigation/Navigation2D node was available but this is no longer always the case with the addition of the default navigation map of the world resource so a user toggle was required (and the 4.0 GridMap had this toggle anyway). For compatibility this toggle is automatically enabled when a Navigation/Navigation2D node is found.

**NavigationLayers**
This will largely affect 4.0 and is not an issue for 3.5 as the navigation layer feature was not available before. The multiple navigation layer function versions were streamlined to "set_navigation_layers" and "get_navigation_layers" (similar to collision physics) across the board as epecially "set_layers" will become a confusing hazard function when other layer types are added.

### **Test Project**
Here is a quick test project for 3.5 for 2D and 3D to test some of the features e.g. GridMap and merging.
[GodotNavigationDebugTest3.x.zip](https://github.com/godotengine/godot/files/8891900/GodotNavigationDebugTest3.x.zip)

### CODE REVIEW

**Comment by akien-mga** [comment]
CC @AndreaCatania @Scony - the aim is to merge this ASAP to include in 3.5 RC 4 so we can have navigation mostly on par with 4.0 in 3.5.

File: modules/navigation/nav_map.cpp [hunk 1]
```diff
@@ -725,30 +710,38 @@ void NavMap::dispatch_callbacks() {
...
 		}
 	}
 }
+
+NavMap::NavMap() {
+}
+
+NavMap::~NavMap() {
+	step_work_pool.finish();
+}
```
Comment (akien-mga) [inline modules/navigation/nav_map.cpp:747]: Missing newline at EOF.

**Comment by smix8** [comment]
@Scony I guess this is good for testing now. You might want to look especially at the parts that you added to 3.5 that I changed or removed as sometimes I couldn't figure out what to do with it cause the code was so different with 4.0. Maybe some of it could and should be reimplemented.

**Review by akien-mga — APPROVED** [review]
Looks pretty good! Let's give this a spin in the next RC and hopefully everything sails smoothly :)

**Comment by akien-mga** [comment]
Thanks!

**Comment by nklbdev** [comment]
It is amazing!

**Pull request merged**